### PR TITLE
Refactor/http gateway

### LINKED ISSUE #1150: Gateway broken for IPNS entries
Go to http://localhost:8080/ipns/QmSJ5B27yQe9bDLmwqF4FxJpBBcpxMnK9uJwX8qbQUyycN

Result:
```
Path Resolve error: multihash length inconsistent: &{124  130 [210]}
```

`ipfs name resolve QmSJ5B27yQe9bDLmwqF4FxJpBBcpxMnK9uJwX8qbQUyycN` on the other hand works fine.

@whyrusleeping

### PR DESCRIPTION
**Pull request by cryptix**
This is supposed to fix #1150 and clean up the gateway http handler in general. Based on #1189.

This is currently WIP but I need some feedback first, see my `TODO`comments in the code.

### CODE REVIEW

File: core/commands/object.go [hunk 1]
```diff
@@ -344,39 +344,21 @@ Data should be in the format specified by the --inputenc flag.
 }
 
 // objectData takes a key string and writes out the raw bytes of that node (if there is one)
-func objectData(n *core.IpfsNode, fpath path.Path) (io.Reader, error) {
-	dagnode, err := core.Resolve(n, fpath)
+func objectData(n *dag.Node, err error) (io.Reader, error) {
```
Comment (wking) [inline core/commands/object.go]: There's so little left in this function after you remove the `Resolve` call that I'd just drop this (and `objectLinks`, etc.) entirely.  Unless we really want to keep them around for the debug logging…

File: core/corehttp/gateway_handler.go [hunk 2]
```diff
@@ -75,39 +68,7 @@ func (i *gatewayHandler) loadTemplate() error {
...
-
-	node, err := i.node.Resolver.ResolvePath(path.Path(p))
-	if err != nil {
-		return nil, "", err
-	}
-	return node, p, err
-}
-
+// TODO(cryptix):  these four helper funcs shoudl also be available elsewhere, i think?
 func (i *gatewayHandler) NewDagFromReader(r io.Reader) (*dag.Node, error) {
```
Comment (wking) [inline core/corehttp/gateway_handler.go]: There's the stuff in `importer` that this method is using.  I guess this is just for the convenience of binding it to the `gatewayHandler`?

**Comment by wking** [comment]
On Sat, May 02, 2015 at 08:07:34PM -0700, Henry wrote:
>   * core: add context.Context param to core.Resolve()

So I've been thinking about this a bit, and I can't wrap my head
around Contexts.  There are lots of calls in Go's standard libraries
that *don't* take Context calls.  Are they all non-blocking?  Where do
you draw the line for “we better accept a Context in case the caller
wants to bail”?  Is this just for stuff that will eventually get wired
up to a web server where the caller could drop the TCP request?  Maybe
we just want a goroutine per request and we can kill the whole
goroutine?  Does Go do single-process event looping like Nginx workers
[1] and Python's asyncio [2]?  Is there no more elegant way to stop
processing for a particular request without passing around Context
parameters?

On the other hand, I barely speak Go, so feel free to ignore my
confusion and just carry on writing idiomatic Go (whatever that
happens to be ;).

[1]: http://nginx.org/en/docs/beginners_guide.html
[2]: https://docs.python.org/3/library/asyncio.html

File: core/corehttp/gateway_handler.go [hunk 3]
```diff
@@ -162,7 +116,7 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		errmsg = errmsg + "bad request for " + r.URL.Path
 	}
 	w.Write([]byte(errmsg))
-	log.Debug(errmsg)
+	log.Error(errmsg) // TODO(cryptix): Why are we ignoring handler errors?
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: we should never ignore errors. any time an error is ignored anywhere in our code, its a bug.

File: core/corehttp/gateway_handler.go [hunk 4]
```diff
@@ -28,13 +28,6 @@ const (
 	IpnsPathPrefix = "/ipns/"
 )
 
-type gateway interface {
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: private interface? yeah, get this outta here.

File: core/commands/object.go (hunk 1 above)
Comment (whyrusleeping) [inline core/commands/object.go]: i think its probably worth dropping, it provides no tangible value added.

File: core/commands/publish.go [hunk 5]
```diff
@@ -89,7 +91,8 @@ Publish an <ipfs-path> to another public key (not implemented):
 		}
 
 		// TODO n.Keychain.Get(name).PrivKey
-		output, err := publish(n, n.PrivateKey, p)
+		// TODO(cryptix): is req.Context().Context a child of n.Context()?
```
Comment (whyrusleeping) [inline core/commands/publish.go]: I think that it is part of the nodes 'context group' somehow. Although I dont claim to understand context groups, or how they different from processes

**Comment by whyrusleeping** [comment]
@wking contexts should be used when you are going to either perform RPC's, spawn off child goroutines you want to handle, or when you want to be able to cancel an operation. Not everything needs to take them, but our high level things should take them, otherwise things may hang. Contexts also allow us to more easily clean up after ourselves; cancelling a context says to all your children "stop whatever you are doing, clean up, and exit ASAP"

**Comment by wking** [comment]
On Sat, May 02, 2015 at 08:51:09PM -0700, Jeromy Johnson wrote:
> @wking contexts should be used when you are going to either perform
> RPC's, spawn off child goroutines you want to handle, or when you
> want to be able to cancel an operation. Not everything needs to take
> them, but our high level things should take them, otherwise things
> may hang.

Lots of things can block besides RPCs and goroutines.  Why doesn't
io.Reader.ReadAll() take a Context?  Or sync.Mutex.Lock()?

> Contexts also allow us to more easily clean up after ourselves;
> cancelling a context says to all your children "stop whatever you
> are doing, clean up, and exit ASAP"

Which you can do between processes with signals, or in an event-loop
situation just by releasing the state associated with the aborted
request.  Are Contexts really the easiest way to do this in Go?

**Comment by whyrusleeping** [comment]
Ill let google explain it better than I can, lol https://blog.golang.org/context

File: core/commands/object.go (hunk 1 above)
Comment (cryptix) [inline core/commands/object.go]: Thought so too, this is mixed up from PR#1189 though. I'll remove it there and rebase this once it is merged.

**Comment by jbenet** [comment]
@wking on the contexts-- the reality with Go is there's no hard and fast "right" way to do all these things. contexts are very nice in many cases, and very hairy in others (try mixing it with net.Conn's, io, and so on). As with most things, Go tries to cover 80% of the cases really simply, and does a very good job at this. We'll certainly find corner cases where things get complex (and even annoying), but for the most part contexts are a very elegant solution to a difficult problem. Anyway, yeah-- read up on it. and read up on the various takes on it. Take a look at how the examples Sameer gives work(and perhaps some of the more esoteric cases, like the go grpc implementation).

File: core/commands/publish.go (hunk 5 above)
Comment (jbenet) [inline core/commands/publish.go]: ctxgroup is old and should be removed. ctxgroup should be replaced by goprocess. the only reason i haven't done it is that goprocess isn't perfect yet-- i believe it has a memleak atm. :( 

if we want to abandon goprocess and ctxgroup for something even simpler, down to think about it. but the problems are pretty annoying -- many cancellations, want semantics where you know the async op has completed, want to query (without blocking) whether something is closed or not, want to wait on things to finish, etc. This problem is not perfectly solved by anything out there-- ctxgroup -> goprocess is _an_ attempt at it. (I know there is a similar internal package at google).

File: core/corehttp/gateway_handler.go [hunk 6]
```diff
@@ -75,39 +68,7 @@ func (i *gatewayHandler) loadTemplate() error {
...
-	}
-
-	node, err := i.node.Resolver.ResolvePath(path.Path(p))
-	if err != nil {
-		return nil, "", err
-	}
-	return node, p, err
-}
-
+// TODO(cryptix):  these four helper funcs shoudl also be available elsewhere, i think?
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: agreed.

File: core/corehttp/gateway_handler.go (hunk 3 above)
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: > any time an error is ignored anywhere in our code, its a bug.

very much agreed.

though:

1. logging is not always an ignored error. sometimes there is nothing to do in a messy io environment (e.g. receiving incoming connections from things that are not running the ipfs protocol. there is nothing to do but log it).
2. this error isn't ignored. an HTTP error header is written, and the error itself is written out to the user for feedback. we could do something else for bookkeeping (like add it to a counter somewhere, throw an event, etc), but correctness is ok here.

**Comment by cryptix** [comment]
TODO:

- [ ] Reenable `t0111-writable.sh` if #1189 is merged as is

**Comment by jbenet** [comment]
@cryptix I'll rebase this on master so the CR is easier.

File: core/corehttp/gateway_handler.go [hunk 7]
```diff
@@ -76,80 +68,31 @@ func (i *gatewayHandler) loadTemplate() error {
...
+			i.putHandler(w, r)
+			return
+		case "DELETE":
+			i.deleteHandler(w, r)
+			return
+		}
 	}
 
-	if r.Method == "HEAD" {
+	if r.Method == "GET" || r.Method == "HEAD" {
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: why not `case "GET", "HEAD":` in the switch?

File: core/corehttp/gateway_handler.go [hunk 8]
```diff
@@ -316,16 +259,16 @@ func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
...
-	if urlPath == IpfsPathPrefix+"QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn/" {
+	if urlPath == ipfsPathPrefix+"QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn/" {
 		i.putEmptyDirHandler(w, r)
 		return
 	}
 
 	var newnode *dag.Node
 	if pathext[len(pathext)-1] == '/' {
-		newnode = NewDagEmptyDir()
+		newnode = uio.NewDirectory(i.node.DAG).GetNode()
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: I think `NewDagEmptyDir()` is more clear and less error prone.

File: core/corehttp/gateway_handler.go [hunk 9]
```diff
@@ -366,19 +314,19 @@ func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
...
 	// resolving path components into merkledag nodes. if a component does not
 	// resolve, create empty directories (which will be linked and populated below.)
-	path_nodes, err := i.node.Resolver.ResolveLinks(tctx, rootnd, components[:len(components)-1])
+	pathNodes, err := i.node.Resolver.ResolveLinks(tctx, rootnd, components[:len(components)-1])
 	if _, ok := err.(path.ErrNoLink); ok {
 		// Create empty directories, links will be made further down the code
-		for len(path_nodes) < len(components) {
-			path_nodes = append(path_nodes, NewDagEmptyDir())
+		for len(pathNodes) < len(components) {
+			pathNodes = append(pathNodes, uio.NewDirectory(i.node.DAG).GetNode())
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: yeah `NewDagEmptyDir()` would def be more clear here

File: core/corehttp/gateway_handler.go [hunk 10]
```diff
@@ -399,7 +347,7 @@ func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	w.Header().Set("IPFS-Hash", key.String())
-	http.Redirect(w, r, IpfsPathPrefix+key.String()+"/"+strings.Join(components, "/"), http.StatusCreated)
+	http.Redirect(w, r, ipfsPathPrefix+key.String()+"/"+strings.Join(components, "/"), http.StatusCreated)
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: these sort of lines are an error-prone mess, maybe we can do a little bit of cleanup?

```go
pcmp = append([]string{"ipfs", key.String()}, components...)
path.FromSegments(pcmp)
```

File: core/corehttp/gateway_handler.go [hunk 11]
```diff
@@ -463,7 +418,7 @@ func (i *gatewayHandler) deleteHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	w.Header().Set("IPFS-Hash", key.String())
-	http.Redirect(w, r, IpfsPathPrefix+key.String()+"/"+strings.Join(components[:len(components)-1], "/"), http.StatusCreated)
+	http.Redirect(w, r, ipfsPathPrefix+key.String()+"/"+strings.Join(components[:len(components)-1], "/"), http.StatusCreated)
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: more error prone-ness

**Comment by jbenet** [comment]
@cryptix #1189 was merged, so re-enable?

File: core/corehttp/gateway_handler.go (hunk 7 above)
Comment (cryptix) [inline core/corehttp/gateway_handler.go]: the above only is for `Writeable`. I didn't want to introduce another switch with just one case.

Comment (jbenet) [inline core/corehttp/gateway_handler.go]: ohhh missed that-- my mistake :)

File: core/corehttp/gateway_handler.go (hunk 10 above)
Comment (wking) [inline core/corehttp/gateway_handler.go]: On Fri, May 08, 2015 at 01:39:50AM -0700, Juan Batiz-Benet wrote:
> > -	http.Redirect(w, r, IpfsPathPrefix+key.String()+"/"+strings.Join(components, "/"), http.StatusCreated)
> > +	http.Redirect(w, r, ipfsPathPrefix+key.String()+"/"+strings.Join(components, "/"), http.StatusCreated)
> 
> these sort of lines are an error-prone mess, maybe we can do a little bit of cleanup?
> 
> ```go
> pcmp = append([]string{"ipfs", key.String()}, components...)
> path.FromSegments(pcmp)

I'm tweaking FromSegments to take an explicit prefix in #1208 [1] (to
support /dns/… inputs, etc.).  With that form, this becomes:

  pcmp = append([]string{key.String()}, components...)
  path.FromSegments("/ipfs/", pcmp...)

or, if we don't want to hard-code /ipfs/ (I don't mind either way):

  path.FromSegments(ipfsPathPrefix, pcmp...)

[1]: https://github.com/wking/go-ipfs/commit/e7dc44876c73c9a9cc1322a0a7e951935f77ff10#diff-48371ba6757afda8e5a57b6dccac57f6R47

**Comment by cryptix** [comment]
rebased

**Comment by jbenet** [comment]
@cryptix may want to squash  1a7b85b wherever it's supposed to go, else some previous commit wont pass all tests

**Comment by cryptix** [comment]
@jbenet Yea, that one was stupid..  i'll squash related once when I'm done.

**Comment by jbenet** [comment]
#1219 merged the changed without the put/writable tests parts. Rebased on top of new master.

**Pull request merged**